acquia/mc-cs-plugin-custom-objects
# Maut 3961 - Improve Link Items UI

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://backlog.acquia.com/browse/MAUT-3961
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
See https://backlog.acquia.com/browse/MAUT-3961

#### Steps to test this PR:
_dev note_: checkout https://github.com/mautic-inc/mautic-cloud/pull/916 in mautic-cloud repo.

1. Create some custom items so you can link them to a custom item or contact.
2. Go to a contact detail page and click on some custom item tab.
3. You should no longer see a search box `Search for a Product to link it with this contact`. There is a new button `Link existing` instead.
4. Click that button and a modal window with table containing unlinked custom items appears.
5. Click a dropdown menu at a given row and select `Link`.
6. The given row should disappear from the modal table and appear in the detail table.
7. Go to a custom item detail page and click on some custom item tab. Repeat the steps 3-6.
8. Test when there is a custom item with parent/child object linking to a contact (there is no such functionality when linking a custom item to another custom item)
    8.1. See https://github.com/mautic-inc/plugin-custom-objects/pull/204 on how to create custom items with parent/child relationship
    8.2. Link such a custom item to a contact.
    8.3. Now when clicking a `Link` action a modal window `New Detail` should appear instead.
    8.4. Clicking the close icon in the modal's top right corner should get you back to the modal window displaying the table.
    8.5. Whereas clicking the `Save & Close` button should get you back to the modal window with the table as well but the given row should disappear and appear in the detail table.

#### Other areas of Mautic that may be affected by the change:
1. Custom item listing.
2. Custom item listing on contact/custom item detail pages.
3. Adding a campaign action `Link contact with <custom item name>`
